va/va.c

```diff
@@ -114,75 +114,100 @@ int vaDisplayIsValid(VADisplay dpy)
 }
 
-static void default_log_error(const char *buffer)
+/*
+ * Global log level configured from the config file or environment, which sets
+ * whether default logging appears or not (always overridden by explicitly
+ * user-configured logging).
+ */
+static int default_log_level = 2;
+
+static void default_log_error(void *user_context, const char *buffer)
 {
+    if (default_log_level < 1)
+        return;
 # ifdef ANDROID
     ALOGE("%s", buffer);
 # else
     fprintf(stderr, "libva error: %s", buffer);
 # endif
 }
 
-static void default_log_info(const char *buffer)
+static void default_log_info(void *user_context, const char *buffer)
 {
+    if (default_log_level < 2)
+        return;
 # ifdef ANDROID
     ALOGI("%s", buffer);
 # else
     fprintf(stderr, "libva info: %s", buffer);
 # endif
 }
 
-static vaMessageCallback va_log_error = default_log_error;
-static vaMessageCallback va_log_info = default_log_info;
-
 /**
  * Set the callback for error messages, or NULL for no logging.
  * Returns the previous one, or NULL if it was disabled.
  */
-vaMessageCallback vaSetErrorCallback(vaMessageCallback callback)
+VAMessageCallback vaSetErrorCallback(VADisplay dpy, VAMessageCallback callback, void *user_context)
 {
-    vaMessageCallback old_callback = va_log_error;
-    va_log_error = callback;
+    VADisplayContextP dctx;
+    VAMessageCallback old_callback;
+
+    if (!vaDisplayIsValid(dpy))
+        return NULL;
+
+    dctx = (VADisplayContextP)dpy;
+    old_callback = dctx->error_callback;
+
+    dctx->error_callback = callback;
+    dctx->error_callback_user_context = user_context;
+
     return old_callback;
 }
 
 /**
  * Set the callback for info messages, or NULL for no logging.
  * Returns the previous one, or NULL if it was disabled.
  */
-vaMessageCallback vaSetInfoCallback(vaMessageCallback callback)
+VAMessageCallback vaSetInfoCallback(VADisplay dpy, VAMessageCallback callback, void *user_context)
 {
-    vaMessageCallback old_callback = va_log_info;
-    va_log_info = callback;
+    VADisplayContextP dctx;
+    VAMessageCallback old_callback;
+
+    if (!vaDisplayIsValid(dpy))
+        return NULL;
+
+    dctx = (VADisplayContextP)dpy;
+    old_callback = dctx->error_callback;
+
+    dctx->info_callback = callback;
+    dctx->info_callback_user_context = user_context;
+
     return old_callback;
 }
 
-void va_MessagingInit()
+static void va_MessagingInit()
 {
 #if ENABLE_VA_MESSAGING
     char env_value[1024];
+    int ret;
 
     if (va_parseConfig("LIBVA_MESSAGING_LEVEL", &env_value[0]) == 0) {
-        if (strcmp(env_value, "0") == 0) {
-            vaSetInfoCallback(NULL);
-            vaSetErrorCallback(NULL);
-        }
-
-        if (strcmp(env_value, "1") == 0) {
-            vaSetInfoCallback(NULL);
-        }
+        ret = sscanf(env_value, "%d", &default_log_level);
+        if (ret < 1)
+            default_log_level = 2;
     }
```

Comment:
It would be better to limit default_log_level to 0, 1, 2 when user set other values in the config file or environment.